src/java.base/share/classes/com/sun/crypto/provider/DHKEM.java

```diff
@@ -0,0 +1,396 @@
+                        "sk".getBytes(StandardCharsets.UTF_8),
+                        "".getBytes(StandardCharsets.UTF_8),
+                        Nsk);
+                PrivateKey k = DeserializePrivateKey(sk);
+                return new KeyPair(getPublicKey(k), k);
+            }
+        }
+
+        private PrivateKey DeserializePrivateKey(byte[] data) throws Exception {
+            KeySpec keySpec = isEC()
+                    ? new ECPrivateKeySpec(new BigInteger(1, (data)), (NamedCurve) spec)
+                    : new XECPrivateKeySpec(spec, data);
+            return KeyFactory.getInstance(keyAlgorithm).generatePrivate(keySpec);
+        }
+    }
+
+    private static SecureRandom getSecureRandom(SecureRandom userSR) {
+        return userSR != null ? userSR : JCAUtil.getSecureRandom();
+    }
+
+    @Override
+    public EncapsulatorSpi engineNewEncapsulator(
+            PublicKey pk, AlgorithmParameterSpec spec, SecureRandom secureRandom)
+            throws InvalidAlgorithmParameterException, InvalidKeyException {
+        if (pk == null) {
+            throw new InvalidKeyException("input key is null");
+        }
+        if (spec != null) {
+            throw new InvalidAlgorithmParameterException("no spec needed");
+        }
+        Params params = paramsFromKey(pk);
+        return new Handler(params, getSecureRandom(secureRandom), null, pk);
+    }
+
+    @Override
+    public DecapsulatorSpi engineNewDecapsulator(PrivateKey sk, AlgorithmParameterSpec spec)
+            throws InvalidAlgorithmParameterException, InvalidKeyException {
+        if (sk == null) {
+            throw new InvalidKeyException("input key is null");
+        }
+        if (spec != null) {
+            throw new InvalidAlgorithmParameterException("no spec needed");
+        }
+        Params params = paramsFromKey(sk);
+        return new Handler(params, null, sk, params.getPublicKey(sk));
+    }
+
+    private Params paramsFromKey(Key k) throws InvalidKeyException {
+        if (k instanceof ECKey eckey) {
+            if (ECUtil.equals(eckey.getParams(), CurveDB.lookup("secp256r1"))) {
+                return Params.P256;
+            } else if (ECUtil.equals(eckey.getParams(), CurveDB.lookup("secp384r1"))) {
+                return Params.P384;
+            } else if (ECUtil.equals(eckey.getParams(), CurveDB.lookup("secp521r1"))) {
+                return Params.P521;
+            }
+        } else if (k instanceof XECKey xkey
+                && xkey.getParams() instanceof NamedParameterSpec ns) {
+            if (ns.getName().equals("X25519")) {
+                return Params.X25519;
+            } else if (ns.getName().equals("X448")) {
+                return Params.X448;
+            }
+        }
+        throw new InvalidKeyException("Unsupported key");
+    }
+
+    private static byte[] concat(byte[]... inputs) {
+        ByteArrayOutputStream o = new ByteArrayOutputStream();
+        Arrays.stream(inputs).forEach(o::writeBytes);
```

Comment:
Unless I'm missing something there is no `stream(byte[])` support, so I'm not sure how this is compiling.  I didn't think  the generics would work with this.
More importantly, `forEach()` the API states that stream is in a non-deterministic order.  I think you want `forEachOrdered()`

Reply:
Ok, this is fine, I didn't realize sequential streams were default